Add vectorsearch Remote Index Build docs (#9575)
opensearch-project/documentation-website: _vector-search/remote-index-build.md

---
layout: default
title: Remote index build
nav_order: 72
has_children: false
redirect_from:
  - /search-plugins/knn/remote-index-build/
---

# Building vector indexes remotely with GPUs
This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/k-NN/issues/2391).    
{: .warning}

> As of version 3.0.0, OpenSearch supports the ability to build vector indexes using GPUs with a remote index build service. This new feature leverages the power of GPUs to dramatically reduce index build times and decrease cost. See the [GPU Accelerated Vector Search blog post](https://opensearch.org/blog/GPU-Accelerated-Vector-Search-OpenSearch-New-Frontier/) for more details.

Reviewer: Does "with a remote index build service" mean "hosted on a remote index build service"?

Author: Technically it's the remote build service that is hosted on GPU machines. "using GPUs **via** a remote build service" may not be the best phrasing, but that is semantically what I am trying to express.

## Getting started

Before configuring the remote index build settings, there are a couple prerequisite steps:

### Step 1: Enable remote index build

In order to use the remote build feature, it must be enabled both for the cluster and the chosen index:

Setting | Static/Dynamic | Default | Description
:--- | :--- | :--- | :---

> `knn.feature.remote_index_build.enabled` | Dynamic | `false` | Feature flag that enables remote vector index building functionality for the cluster.

Reviewer: This is the experimental flag, correct? It does not have `experimental` in the flag name, like other experimental flags?

Reviewer 2: This is the convention KNN feature flags follow: https://github.com/opensearch-project/k-NN/blob/3a48c4b31599e42eccf8b9015d5bd78c5c31aac2/src/main/java/org/opensearch/knn/common/featureflags/KNNFeatureFlags.java#L28-L29

If the feature flag naming is an issue we can take a follow up action to unify it (although in general, feature flags don't really work from plugins. Ref: https://github.com/opensearch-project/OpenSearch/pull/17952)

Reviewer: Not an issue, thanks for the details!

Author: @jed326 In that case, [Experimental feature flags](https://docs.opensearch.org/docs/latest/install-and-configure/configuring-opensearch/experimental/) doesn't apply to this setting, right? In my experience, I've only set it using the cluster settings API.

Reviewer 2: Yes. You can refer to the PR in my previous comment for why that does not work for plugins currently.

Author: Thanks. @kolchfa-aws I think we can still refer to it as a feature flag although the link provided may cause confusion, since it will be set in the same fashion as the other settings listed. What do you think?

Reviewer: If it can only be set as a dynamic setting, let's not refer to it as a feature flag and remove the link. I'll make the change.

> `index.knn.remote_index_build.enabled` | Dynamic | `false` | Whether remote index building is enabled for this index. Currently, the remote index build service supports Faiss indexes with method HNSW and data type `FP32`.

Reviewer: This is a regular dynamic setting that users can set either through the API or in the yaml file?

Author: Yes.

> `index.knn.remote_index_build.enabled` | Dynamic | `false` | Whether remote index building is enabled for this index. Currently, the remote index build service supports Faiss indexes with method HNSW and data type `FP32`.

Reviewer: FP32 means 32-bit floating-point values? So default (with no quantization)?

Author: Yes, it should be clear that the remote index build service will not build indexes of vectors that are binary or byte data types or quantized.

### Step 2: Create and register the remote vector repository

> The remote vector repository will act as an intermediate object store between the OpenSearch cluster and the remote build service. The cluster will upload vectors and doc IDs to this repository. The remote build service will download these and eventually upload a completed index build in response. To create and register the repository, follow the [Register repository](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/#register-repository) steps in the Snapshots documentation. Then, set  `knn.remote_index_build.vector_repo` to be the name of this registered repository.

Reviewer: Is `knn.remote_index_build.vector_repo` a dynamic setting that users can update via the API or yaml file?

Reviewer: Does "completed index build" mean "built index"? Does "in response" mean that the build service uploads the finished (built) index as a result of processing the uploaded vectors and document IDs?

Author: 1. Yes, a dynamic setting as per the definition in the "Dynamic settings" link you provided above.
2. "Built index" is better here, thanks.
3. Yes. It uploads the built index once it finishes the operation.

> The remote build service feature currently only supports Amazon S3 repositories.

Reviewer 2: So from k-nn perspective we can support any of the vendor implementations, but the remote vector index builder we provide only supports S3. So I think we should move this down to the next section.

Author: Currently the code (in k-NN) will throw an exception for any repository type other than `s3`. It's not just a `remote-vector-index-builder` check. I think a user following this guide should know this when creating their repository. If we want to keep k-NN open to different vendors we should remove that check ([ref](https://github.com/owenhalpert/k-NN/blob/main/src/main/java/org/opensearch/knn/index/codec/nativeindex/remote/RemoteIndexBuildStrategy.java#L316-L319)), although I see it in the same vein as checking for Faiss, certain space types, etc, which are all based on the `remote-vector-index-builder` capabilities.

---
layout: default
title: Remote index build
nav_order: 72
has_children: false
---

# Building vector indexes remotely using GPUs
Introduced 3.0 
{: .label .label-purple }

This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/k-NN/issues/2391).    
{: .warning}

> Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. 

Reviewer: ```suggestion
Starting with version 3.0, OpenSearch supports building vector indexes using a GPU-accelerated remote index build service. Using GPUs dramatically reduces index build times and decreases costs. 
```

> Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. 

Reviewer 2: @kolchfa-aws part of the information we want to share via the blog post link is the preliminary performance benchmarking results. That doesn't seem like something we would include in this page, so what would be the correct way to link to those results?

> Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. 

Reviewer: Makes sense. We can link to the benchmarking section of the blog directly.

> Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. 

Reviewer 2: I'm not sure how to link to a specific section, if you do could you help link to the `Benchmark results` section of  https://opensearch.org/blog/GPU-Accelerated-Vector-Search-OpenSearch-New-Frontier/? Thanks!

> Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. 

Reviewer: ```suggestion
Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. For benchmarking results, see [this blog post](https://opensearch.org/blog/GPU-Accelerated-Vector-Search-OpenSearch-New-Frontier/).
```

> Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. 

Reviewer: Turns out we can't link to a section of the blog. I added a link with the text mentioning benchmarks.

> Starting with version 3.0, OpenSearch supports building vector indexes using GPUs with a remote index build service. Using GPUs dramatically reduces index build times and decreases costs. For benchmarking results, see [this blog post](https://opensearch.org/blog/GPU-Accelerated-Vector-Search-OpenSearch-New-Frontier/).

Reviewer: ```suggestion
Starting with version 3.0, OpenSearch supports building vector indexes using a GPU-accelerated remote index build service. Using GPUs dramatically reduces index build times and decreases costs. For benchmarking results, see [this blog post](https://opensearch.org/blog/GPU-Accelerated-Vector-Search-OpenSearch-New-Frontier/).
```

## Prerequisites

Before configuring the remote index build settings, ensure you fulfill the following prerequisites. For more information about updating dynamic settings, see [Dynamic settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/index/#dynamic-settings).

### Step 1: Enable remote index build

> Enable remote index build for both for the cluster and the chosen index by configuring the following settings. 

Reviewer: ```suggestion
Enable remote index build for both the cluster and the chosen index by configuring the following settings. 
```

Setting | Static/Dynamic | Default | Description
:--- | :--- | :--- | :---
`knn.feature.remote_index_build.enabled` | Dynamic | `false` | A feature flag that enables remote vector index building for the cluster. For more information about enabling feature flags, see [Experimental feature flags]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/experimental/).
`index.knn.remote_index_build.enabled` | Dynamic | `false` | Enables remote index building for the index. Currently, the remote index build service supports [Faiss]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/#faiss-engine) indexes with the `hnsw` method and the default 32-bit floating-point (`FP32`) vectors. For more information about updating dynamic settings, see [Dynamic settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/index/#dynamic-settings).

### Step 2: Create and register the remote vector repository

> The remote vector repository acts as an intermediate object store between the OpenSearch cluster and the remote build service. The cluster uploads vectors and doc IDs to this repository. The remote build service then downloads these and eventually uploads a completed index build in response. To create and register the repository, follow the steps in [Register repository]({{site.url}}{{site.baseurl}}/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/#register-repository). Then set the `knn.remote_index_build.vector_repo` dynamic setting to the name of this registered repository.

Reviewer: ```suggestion
The remote vector repository acts as an intermediate object store between the OpenSearch cluster and the remote build service. The cluster uploads vectors and document IDs to the repository. The remote build service retrieves the data, builds the index externally, and uploads the completed result back to the repository.

To create and register the repository, follow the steps in [Register repository]({{site.url}}{{site.baseurl}}/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/#register-repository). Then set the `knn.remote_index_build.vector_repo` dynamic setting to the name of this registered repository.
```

---
layout: default
title: Remote index build
nav_order: 72
has_children: false
---

# Building vector indexes remotely using GPUs
Introduced 3.0 
{: .label .label-purple }

This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/k-NN/issues/2391).    
{: .warning}

Starting with version 3.0, OpenSearch supports building vector indexes using a GPU-accelerated remote index build service. Using GPUs dramatically reduces index build times and decreases costs. For benchmarking results, see [this blog post](https://opensearch.org/blog/GPU-Accelerated-Vector-Search-OpenSearch-New-Frontier/).

## Prerequisites

Before configuring the remote index build settings, ensure you fulfill the following prerequisites. For more information about updating dynamic settings, see [Dynamic settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/index/#dynamic-settings).

> ### Step 1: Enable remote index build

Reviewer 3: ```suggestion
### Step 1: Enable the remote index build service
```

> Enable remote index build for both the cluster and the chosen index by configuring the following settings. 

Reviewer 3: ```suggestion
Enable the remote index build service for both the cluster and the chosen index by configuring the following settings. 
```

Setting | Static/Dynamic | Default | Description
:--- | :--- | :--- | :---
`knn.feature.remote_index_build.enabled` | Dynamic | `false` | Enables remote vector index building for the cluster. 
`index.knn.remote_index_build.enabled` | Dynamic | `false` | Enables remote index building for the index. Currently, the remote index build service supports [Faiss]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/#faiss-engine) indexes with the `hnsw` method and the default 32-bit floating-point (`FP32`) vectors. 

### Step 2: Create and register the remote vector repository

The remote vector repository acts as an intermediate object store between the OpenSearch cluster and the remote build service. The cluster uploads vectors and document IDs to the repository. The remote build service retrieves the data, builds the index externally, and uploads the completed result back to the repository.

> To create and register the repository, follow the steps in [Register repository]({{site.url}}{{site.baseurl}}/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/#register-repository). Then set the `knn.remote_index_build.vector_repo` dynamic setting to the name of this registered repository.

Reviewer 3: ```suggestion
To create and register the repository, follow the steps in [Register repository]({{site.url}}{{site.baseurl}}/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/#register-repository). Then set the `knn.remote_index_build.vector_repo` dynamic setting to the name of the registered repository.
```

> The remote build service currently only supports Amazon S3 repositories.

Reviewer 3: ```suggestion
The remote build service currently only supports Amazon Simple Storage Service (Amazon S3) repositories.
```

{: .note}

### Step 3: Set up a remote vector index builder

> Configure the remote endpoint in k-NN settings by setting `knn.remote_index_build.client.endpoint` to a running [remote vector index builder](https://github.com/opensearch-project/remote-vector-index-builder) instance. For instructions to set up the remote service, see [the user guide](https://github.com/opensearch-project/remote-vector-index-builder/blob/main/USER_GUIDE.md).

Reviewer 3: ```suggestion
Configure the remote endpoint in the k-NN settings by setting `knn.remote_index_build.client.endpoint` to a running [remote vector index builder](https://github.com/opensearch-project/remote-vector-index-builder) instance. For instructions on setting up the remote service, see [the user guide](https://github.com/opensearch-project/remote-vector-index-builder/blob/main/USER_GUIDE.md).
```

## Configuring remote index build settings

> Remote index build supports several additional optional settings. For information about configuring any remaining remote index build settings, see [Remote index build settings]({{site.url}}{{site.baseurl}}/vector-search/settings/#remote-index-build-settings).

Reviewer 3: ```suggestion
The remote index build service supports several additional, optional settings. For information about configuring any remaining remote index build settings, see [Remote index build settings]({{site.url}}{{site.baseurl}}/vector-search/settings/#remote-index-build-settings).
```

> ## Using remote index build

Reviewer 3: ```suggestion
## Using the remote index build service
```

> Once remote index build is configured, any index with it enabled will use the remote vector index builder for builds that meet the configured `index.knn.remote_index_build.size_threshold`.

Reviewer 3: ```suggestion
Once the remote index build service is configured, any index on which it is enabled will use the remote vector index builder for builds that meet the configured `index.knn.remote_index_build.size_threshold`.
```
